### DISCUSSION TITLE
About the method of converting astiav.Frame to Image

### DISCUSSION THREAD

**Discussion by Tryanks** [asticode/go-astiav#30] [Ideas]
pre-reading #4 #5 #22 
A no matter the cost method is to use astiav.Encoder to encode astiav.Frame to PNG and read it as Image.Image.
Is there a better proposal that doesn't require untangling numerous pixel formats?

**Comment by asticode** [comment]
Honestly, now I'd like this part to be part of this lib. The way I see it is:

- create a `FrameData` struct:

```go
type FrameData struct {
    ...
}

func(d *FrameData) Bytes() [NumDataPointers][]byte {}

func(d *FrameData) Image() (image.Image, error) {}

func(d *FrameData) Sample() ("whatever go representation is best suited", error) {}
```

- `Frame.Data()` now returns a `*FrameData` instead of `[NumDataPointers][]byte` directly
- `FrameData.Image()` will check the frame is an image, check whether pixel format is handled and either return `*image.RGBA`, `*image.YCbCr`, `*image.YCbCrA`, ... depending on the pixel format

Does that sound even remotely sound?

If so, I'll have some time this weekend to try it out 👍

**Comment by asticode** [comment]
FYI [Frame.Data()](https://github.com/asticode/go-astiav/blob/master/frame.go#L57) now returns a [*FrameData](https://github.com/asticode/go-astiav/blob/master/frame_data.go#L12) that allows to either get data as `[]byte` or, if it's an image, get data as an `image.Image` 👍

**Reply by Tryanks** [reply]
Great job! I can't believe you actually did it. I tried converting directly based on bytes once, but I gave up after a couple months long learning curve.
